src/ripple/app/tx/impl/NFTokenAcceptOffer.cpp

```diff
@@ -441,7 +492,28 @@ NFTokenAcceptOffer::doApply()
             !isTesSuccess(ret))
             return ret;
 
-        return nft::insertToken(view(), buyer, std::move(tokenAndPage->token));
+        auto const sleBuyer = view().read(keylet::account(buyer));
+        std::uint32_t const buyerOwnerCountBefore =
+            sleBuyer->getFieldU32(sfOwnerCount);
+
+        auto const insertRet =
+            nft::insertToken(view(), buyer, std::move(tokenAndPage->token));
+
+        // if fixNFTokenReserve is enabled, check if the buyer has sufficient
+        // reserve to own a new object, if their OwnerCount changed.
+        //
+        // However, this check is merely for safety and should never trigger,
+        // because, in brokered mode, the buy offer's reserve will be freed up,
+        // which can be taken up by the new NFT again. So buyer is always
+        // guarenteed to have enough reserve.
+        if (view().rules().enabled(fixNFTokenReserve))
+        {
+            if (auto const ret =
+                    checkBuyerReserve(sleBuyer, buyerOwnerCountBefore);
+                !isTesSuccess(ret))
+                return ret;
+        }
+        return insertRet;
```

Comment:
I was a bit bothered that the code inserted into `acceptOffer` and `doApply` was very close to identical.  So I did a bit of nosing about.  It looks to me that the code that actually transfers the token in `acceptOffer` and `doApply` could be a common helper function.  It might look like this:
```
TER
NFTokenAcceptOffer::transferNFToken(
    AccountID const& buyer, AccountID const& seller, uint256 const& nftokenID)
{
    auto tokenAndPage = nft::findTokenAndPage(view(), seller, nftokenID);

    if (!tokenAndPage)
        return tecINTERNAL;

    if (auto const ret = nft::removeToken(
            view(), seller, nftokenID, std::move(tokenAndPage->page));
        !isTesSuccess(ret))
        return ret;

    auto const sleBuyer = view().read(keylet::account(buyer));
    if (!sleBuyer)
        return tecINTERNAL;

    std::uint32_t const buyerOwnerCountBefore =
        sleBuyer->getFieldU32(sfOwnerCount);

    auto const insertRet =
        nft::insertToken(view(), buyer, std::move(tokenAndPage->token));

    // if fixNFTokenReserve is enabled, check if the buyer has sufficient
    // reserve to own a new object, if their OwnerCount changed.
    //
    // There was an issue where the buyer accepts a sell offer, the ledger
    // didn't check if the buyer has enough reserve, meaning that buyer can get
    // NFTs free of reserve.
    if (view().rules().enabled(fixNFTokenReserve))
    {
        // To check if there is sufficient reserve, we cannot use mPriorBalance
        // because NFT is sold for a price. So we must use the balance after
        // the deduction of the potential offer price. A small caveat here is
        // that the balance has already deducted the transaction fee, meaning
        // that the reserve requirement is a few drops higher.
        auto const buyerBalance = sleBuyer->getFieldAmount(sfBalance);

        auto const buyerOwnerCountAfter = sleBuyer->getFieldU32(sfOwnerCount);
        if (buyerOwnerCountAfter > buyerOwnerCountBefore)
        {
            if (auto const reserve =
                    view().fees().accountReserve(buyerOwnerCountAfter);
                buyerBalance < reserve)
                return tecINSUFFICIENT_RESERVE;
        }
    }

    return insertRet;
}
```
Consider replacing `checkBuyerReserve()` with a function like this.  Then you'll be able to reduce the size of both `acceptOffer` and `doApply` and increase confidence that they are both doing the same thing more often.

If you are curious, you can see an example of what I'm describing as the top-most commit here: https://github.com/scottschurr/rippled/commits/shawnx-nft-reserve

Reply:
Thanks for the suggestion! This looks much better